### PR TITLE
OCPBUGSM-18040 Cluster installation fails with error : the container …

### PR DESCRIPTION
**Pull request by eranco74**
…name \"assisted-installer\" is already in use.

Failure is caused by executing multiple install commands on the host.
Since only one container with the name "assisted-installer" can be executed just one install command is actually running and the rest return with error setting the host installation to failure
The trigger to this behavior is the time it take to pull the installer image from the downstream registry.
This patch mitigate the error by not moving the host to error in case the install command fail with that specific error code and stderr output (the container name \"assisted-installer\" is already in use) .

### CODE REVIEW

File: subsystem/cluster_test.go [hunk 1]
```diff
@@ -781,6 +781,76 @@ var _ = Describe("cluster install", func() {
 
 		})
 
+		It("[only_k8s]install_cluster install command failed", func() {
+			By("Installing cluster till finalize")
```
Comment (filanov) [inline subsystem/cluster_test.go:785]: Don't we already have tests that validate it?

File: subsystem/cluster_test.go [hunk 2]
```diff
@@ -781,6 +781,76 @@ var _ = Describe("cluster install", func() {
...
+					ExitCode: 137,
+					StepType: models.StepTypeInstall,
+					Output:   "blabla",
+					Error:    "Trying to pull registry.stage.redhat.io/openshift4/assisted-installer-rhel8:v4.6.0-19...\nGetting image source signatures\nCopying blob sha256:e5fbed36397a9434b3330d01bcf53befb828e476be291c8e1c026a9753d59dfd\nCopying blob sha256:0fd3b5213a9b4639d32bf2ef6a3d7cc9891c4d8b23639ff7ae99d66ecb490a70\nCopying blob sha256:aebb8c5568533b57ee3da86262f7bff81383a2a624b9f54b9da3418705009901\nCopying config sha256:67bbdf0b8fb27217b9c5f6fa3593925309ef8c95b8b0be8b44713ba5f826fcee\nWriting manifest to image destination\nStoring signatures\nError: error creating container storage: the container name \"assisted-installer\" is already in use by \"331fe687f4af5c7adf75a9ddaaadfb801739ba1815bfcaafd4db7392bf9049bc\". You have to remove that container to be able to reuse that name.: that name is already in use\n",
+					StepID:   string(models.StepTypeInstall),
+				},
+			})
+			Expect(err).ShouldNot(HaveOccurred())
+
+			waitForClusterState(ctx, clusterID, "finalizing", defaultWaitForClusterStateTimeout, "Finalizing cluster installation")
```
Comment (filanov) [inline subsystem/cluster_test.go:842]: please use a const from `models.ClusterStatus...`

File: subsystem/cluster_test.go [hunk 3]
```diff
@@ -781,6 +781,76 @@ var _ = Describe("cluster install", func() {
...
+
+			for _, host := range c.Hosts {
+				updateProgress(*host.ID, clusterID, models.HostStageDone)
+			}
+			// post failure to execute the install command due to a running assisted-installer
+			_, err := agentBMClient.Installer.PostStepReply(ctx, &installer.PostStepReplyParams{
+				ClusterID: clusterID,
+				HostID:    *c.Hosts[0].ID,
+				Reply: &models.StepReply{
+					ExitCode: 137,
```
Comment (filanov) [inline subsystem/cluster_test.go:833]: please add a comment about this specific error (i see the comment above but it's not clear why 137)

File: internal/bminventory/inventory.go [hunk 4]
```diff
@@ -1551,9 +1551,14 @@ func (b *bareMetalInventory) PostStepReply(ctx context.Context, params installer
 	return installer.NewPostStepReplyNoContent()
 }
 
-func handleReplyError(params installer.PostStepReplyParams, b *bareMetalInventory, ctx context.Context, h *models.Host) error {
+func (b *bareMetalInventory) handleReplyError(params installer.PostStepReplyParams, ctx context.Context, h *models.Host) error {
 
 	if params.Reply.StepType == models.StepTypeInstall {
+		// Handle case of installation failure due to an already running assisted-installer.
+		if params.Reply.ExitCode == 137 && strings.Contains(params.Reply.Error, "the container name \"assisted-installer\" is already in use") {
```
Comment (tsorya) [inline internal/bminventory/inventory.go:1558]: Interesting if this will work if logs_sender will fail too as we saw?

Comment (eranco74) [reply internal/bminventory/inventory.go:1558]: That's why I verify this text in the error log:
"the container name \"assisted-installer\" is already in use"

File: subsystem/cluster_test.go (hunk 1 above)
Comment (tsorya) [inline subsystem/cluster_test.go:785]: why we need to go to finalize at all? just fail it on the first state

File: subsystem/cluster_test.go [hunk 5]
```diff
@@ -781,6 +781,76 @@ var _ = Describe("cluster install", func() {
...
+				},
+			})
+			Expect(err).ShouldNot(HaveOccurred())
+
+			By("Verifying installation failed")
+			waitForClusterState(ctx, clusterID, "error", defaultWaitForClusterStateTimeout, "failed")
+		})
+
+		It("[only_k8s]install_cluster assisted-installer already running", func() {
+			By("Installing cluster till finalize")
```
Comment (tsorya) [inline subsystem/cluster_test.go:816]: same here. fail it in installing

File: internal/bminventory/inventory.go (hunk 4 above)
Comment (YuviGold) [inline internal/bminventory/inventory.go:1558]: Make a constant

File: subsystem/cluster_test.go (hunk 3 above)
Comment (YuviGold) [reply subsystem/cluster_test.go:827]: Having a constant variable would make it clearer :)

File: subsystem/cluster_test.go [hunk 6]
```diff
@@ -781,6 +781,73 @@ var _ = Describe("cluster install", func() {
...
+					StepType: models.StepTypeInstall,
+					Output:   "blabla",
+					Error:    "Some random error",
+					StepID:   string(models.StepTypeInstall),
+				},
+			})
+			Expect(err).ShouldNot(HaveOccurred())
+
+			By("Verifying installation failed")
+			waitForClusterState(ctx, clusterID, "error", defaultWaitForClusterStateTimeout, "failed")
```
Comment (filanov) [inline subsystem/cluster_test.go:809]: Please use consts for all the states.

File: subsystem/cluster_test.go [hunk 7]
```diff
@@ -781,6 +781,73 @@ var _ = Describe("cluster install", func() {
 
 		})
 
+		It("[only_k8s]install_cluster install command failed", func() {
+			By("Installing cluster till finalize")
+			c := installCluster(clusterID)
+			Expect(swag.StringValue(c.Status)).Should(Equal("installing"))
```
Comment (filanov) [inline subsystem/cluster_test.go:787]: Please use consts

File: subsystem/cluster_test.go (hunk 1 above)
Comment (filanov) [reply subsystem/cluster_test.go:785]: Looks like we don't

File: subsystem/cluster_test.go [hunk 8]
```diff
@@ -781,6 +781,73 @@ var _ = Describe("cluster install", func() {
...
+			Expect(swag.StringValue(c.StatusInfo)).Should(Equal("Installation in progress"))
+			Expect(len(c.Hosts)).Should(Equal(4))
+			for _, host := range c.Hosts {
+				Expect(swag.StringValue(host.Status)).Should(Equal("installing"))
```
Comment (filanov) [inline subsystem/cluster_test.go:791]: please use consts.

File: internal/bminventory/inventory.go [hunk 9]
```diff
@@ -1551,9 +1552,14 @@ func (b *bareMetalInventory) PostStepReply(ctx context.Context, params installer
 	return installer.NewPostStepReplyNoContent()
 }
 
-func handleReplyError(params installer.PostStepReplyParams, b *bareMetalInventory, ctx context.Context, h *models.Host) error {
+func (b *bareMetalInventory) handleReplyError(params installer.PostStepReplyParams, ctx context.Context, h *models.Host) error {
 
 	if params.Reply.StepType == models.StepTypeInstall {
+		// Handle case of installation error due to an already running assisted-installer.
```
Comment (filanov) [inline internal/bminventory/inventory.go:1558]: 125 is a generic error, what makes it already-running is the combination of the string and the error.
You should probably point it out

**Comment by filanov** [comment]
@eranco74 Looks like both tests that you added failed in jenkins

**Comment by eranco74** [comment]
> @eranco74 Looks like both tests that you added failed in jenkins

Only god knows why we return BadRequest from the PostStepReply API if the exit code in the request isn't 0.
Since we are in code freeze I didn't change the current behavior and updated the test to expect the error.
I'll open a bug for it so we will fix it later

**Comment by filanov** [comment]
> > @eranco74 Looks like both tests that you added failed in jenkins
> 
> Only god knows why we return BadRequest from the PostStepReply API if the exit code in the request isn't 0.
> Since we are in code freeze I didn't change the current behavior and updated the test to expect the error.
> I'll opened a [bug](https://issues.redhat.com/browse/MGMT-2237) for it so we will fix it later

@gamli75 FYI

File: internal/bminventory/inventory.go [hunk 10]
```diff
@@ -65,6 +65,9 @@ const (
 const DefaultUser = "kubeadmin"
 const ConsoleUrlPrefix = "https://console-openshift-console.apps"
 
+// 125 is the generic exit code for cases the error is in podman / docker and not the container we tried to run
+const containerAlreadyRunningExitCode = 125
```
Comment (tsorya) [inline internal/bminventory/inventory.go:69]: didn't we saw 137 too?

Comment (tsorya) [reply internal/bminventory/inventory.go:69]: maybe it is not really interesting to verify error code?

Comment (filanov) [reply internal/bminventory/inventory.go:69]: 125 is a generic error code https://docs.docker.com/engine/reference/run/#exit-status

File: Makefile [hunk 11]
```diff
@@ -24,10 +24,10 @@ ASSISTED_ORG ?= quay.io/ocpmetal
 ASSISTED_TAG ?= latest
 CONTAINER_BUILD_PARAMS = --network=host --label git_revision=${GIT_REVISION} ${CONTAINER_BUILD_EXTRA_PARAMS}
 
-SERVICE := $(or ${SERVICE},${ASSISTED_ORG}/assisted-service:${ASSISTED_TAG})
+export SERVICE := $(or ${SERVICE},${ASSISTED_ORG}/assisted-service:${ASSISTED_TAG})
```
Comment (filanov) [inline Makefile:27]: This change should be tested on PSI and deployment from laptop.

Comment (eranco74) [reply Makefile:27]: This change is required for the subsystem to use the SERVICE specified at the jenkins file and not latest.
@YuviGold I think we should just revert the change that was pushed yesterday.

**Review by YuviGold — APPROVED** [review]

**Review by filanov — APPROVED** [review]

**Pull request merged**